File: chap2/sec2.3/examplePararealHeat.m
f=@(x,t) x.^4.*(1-x).^4+10*sin(8*t);              % heat source function
T=8; N=16; K=16; J=10;                            % parareal parameters
u0=zeros(J+1,1);                                  % initial condition
MG=1;                                             % MG no of coarse steps
gl=zeros(MG+1,1);gr=zeros(MG+1,1);                % G boundary conditions
G=@(t0,t1,u0) SHeatEquationBE(f,[t0 t1],[0 1],u0,gl,gr);  % G coarse solver
MF=10;                                            % MF no of fine steps
gl=zeros(MF+1,1);gr=zeros(MF+1,1);                % F boundary conditions
F=@(t0,t1,u0) SHeatEquationBE(f,[t0 t1],[0 1],u0,gl,gr);  % F fine solver
U=Parareal(F,G,T,u0,N,K);                         
glf=zeros(MF*N+1,1); grf=zeros(MF*N+1,1);         
u=HeatEquationBE(f,[0 T],[0 1],u0,glf,grf);       % fine solution
dt=T/(MF*N);dT=T/N; dx=1/J;                       % mesh parameters
t=(0:dt:T)'; TT=(0:dT:T)'; x=0:dx:1;              % for plotting
for k=1:K
  up=[];
  for n=1:N                                       % reconstruct fine
    up((n-1)*MF+1:n*MF+1,:)=HeatEquationBE(f,...  % solution from 
        [(n-1)*dT n*dT],[0 1],U{k}(n,:),gl,gr);   % parareal for plotting
  end;  
  up(N*MF+1,:)=U{k}(end,:);
  mesh(x,t,up); xlabel('x'); ylabel('t');         % plot parareal approx.
  axis([0 1 0 T -1 1])
  pause
  mesh(x,t,u-up); xlabel('x'); ylabel('t');       % plot parareal error
  err(k)=max(max(abs(u-up)));
  pause
end

File: chap2/sec2.3/examplePararealHeat.py
import numpy as np
import matplotlib.pyplot as plt

from HeatEquationBE import heatEquationBE
from Parareal import parareal

# Problem setup
f = lambda x, t: x**4*(1-x)**4 + 10*np.sin(8*t)  # Heat source function
T = 8
N = 16
K = 16
J = 10
u0 = np.zeros(J + 1)

# Coarse solver
MG = 1
gl = np.zeros(MG + 1)
gr = np.zeros(MG + 1)
G = lambda t0, t1, u0: heatEquationBE(f, (t0, t1), (0, 1), u0, gl, gr)[-1][-1]

# Fine solver
MF = 10
glF = np.zeros(MF + 1)
grF = np.zeros(MF + 1)
F = lambda t0, t1, u0: heatEquationBE(f, (t0, t1), (0, 1), u0, glF, grF)[-1][-1]

# Parareal computation
U = parareal(F, G, T, u0, N, K)

# Fine solution
glf = np.zeros(MF * N + 1)
grf = np.zeros(MF * N + 1)
uFine = heatEquationBE(f, (0, T), (0, 1), u0, glf, grf)[-1]

# Mesh parameters
dt = T / (MF*N)
dT = T / N
dx = 1 / J
t = np.linspace(0, T, MF*N + 1)
TT = np.linspace(0, T, N+1)
x = np.linspace(0, 1, J+1)

# Plotting
fig = plt.figure("examplePararealHeat")
ax = fig.add_subplot(111, projection='3d')

for k in range(K):
    if not plt.fignum_exists("examplePararealHeat"): break

    # Reconstruct fine solution from parareal for plotting
    up = np.zeros_like(uFine)
    for n in range(N):
        up[n*MF:(n+1)*MF+1, :] = heatEquationBE(
            f, (n*dT, (n+1)*dT), (0, 1), U[k, n, :], glF, grF)[-1]
    up[N * MF, :] = U[k, -1, :]

    # Plot parareal approximation
    ax.cla()
    T_grid, X_grid = np.meshgrid(t, x)
    ax.plot_surface(
        X_grid, T_grid, up.T, cmap='viridis',
        rstride=1, cstride=1)
    ax.set_xlabel('x')
    ax.set_ylabel('t')
    ax.set_title(f'Parareal Approximation (Iteration {k})')
    plt.pause(1)

    # Compute and plot error
    error = np.abs(uFine - up)
    ax.cla()
    ax.plot_surface(
        X_grid, T_grid, error.T, cmap='hot',
        rstride=1, cstride=1)
    ax.set_xlabel('x')
    ax.set_ylabel('t')
    ax.set_title(f'Parareal Error (Iteration {k})')
    plt.pause(1)

plt.show()
